### PR TITLE
feat: add logs-dir option

### LINKED ISSUE #4206: [QUESTION] How to disable creating of verbose log files for every npm command?
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

When I run any npm related command I got a new debug log file created, which is kinda annoying.
How can prevent the debug log files being created after each npm command?

Form example when I run `npm -v`, I got a new log file with the following output:
```
0 verbose cli [
0 verbose cli   '/usr/local/Cellar/node/17.3.0/bin/node',
0 verbose cli   '/usr/local/bin/npm',
0 verbose cli   '-v'
0 verbose cli ]
1 info using npm@8.3.0
2 info using node@v17.3.0
3 timing npm:load:whichnode Completed in 0ms
4 timing config:load:defaults Completed in 1ms
5 timing config:load:file:/usr/local/lib/node_modules/npm/npmrc Completed in 3ms
6 timing config:load:builtin Completed in 5ms
7 timing config:load:cli Completed in 3ms
8 timing config:load:env Completed in 0ms
9 timing config:load:project Completed in 3ms
10 timing config:load:file:/Users/(protected)/.npmrc Completed in 1ms
11 timing config:load:user Completed in 2ms
12 timing config:load:file:/usr/local/etc/npmrc Completed in 0ms
13 timing config:load:global Completed in 0ms
14 timing config:load:validate Completed in 0ms
15 timing config:load:credentials Completed in 1ms
16 timing config:load:setEnvs Completed in 1ms
17 timing config:load Completed in 16ms
18 timing npm:load:configload Completed in 17ms
19 timing npm:load:setTitle Completed in 2ms
20 timing config:load:flatten Completed in 2ms
21 timing npm:load:display Completed in 4ms
22 verbose logfile /Users/(protected)/.npm/_logs/2022-01-04T19_56_27_927Z-debug-0.log
23 timing npm:load:logFile Completed in 4ms
24 timing npm:load:timers Completed in 0ms
25 timing npm:load:configScope Completed in 0ms
26 timing npm:load Completed in 27ms
27 verbose exit 0
28 timing npm Completed in 316ms
29 info ok
```

When I start some node app with `npm run dev` (for example, some Nuxt.js app), again I got a new log file:
```
0 verbose cli [
0 verbose cli   '/usr/local/Cellar/node/17.3.0/bin/node',
0 verbose cli   '/usr/local/bin/npm',
0 verbose cli   'run',
0 verbose cli   'dev'
0 verbose cli ]
1 info using npm@8.3.0
2 info using node@v17.3.0
3 timing npm:load:whichnode Completed in 1ms
4 timing config:load:defaults Completed in 2ms
5 timing config:load:file:/usr/local/lib/node_modules/npm/npmrc Completed in 3ms
6 timing config:load:builtin Completed in 4ms
7 timing config:load:cli Completed in 2ms
8 timing config:load:env Completed in 0ms
9 timing config:load:file:/Users/(protected)/.../(protected)/.npmrc Completed in 2ms
10 timing config:load:project Completed in 5ms
11 timing config:load:file:/Users/(protected)/.npmrc Completed in 1ms
12 timing config:load:user Completed in 1ms
13 timing config:load:file:/usr/local/etc/npmrc Completed in 0ms
14 timing config:load:global Completed in 1ms
15 timing config:load:validate Completed in 0ms
16 timing config:load:credentials Completed in 1ms
17 timing config:load:setEnvs Completed in 1ms
18 timing config:load Completed in 17ms
19 timing npm:load:configload Completed in 17ms
20 timing npm:load:setTitle Completed in 3ms
21 timing config:load:flatten Completed in 3ms
22 timing npm:load:display Completed in 4ms
23 verbose logfile /Users/(protected)/.npm/_logs/2022-01-04T18_20_49_151Z-debug-0.log
24 timing npm:load:logFile Completed in 6ms
25 timing npm:load:timers Completed in 0ms
26 timing npm:load:configScope Completed in 0ms
27 timing npm:load Completed in 31ms
```

And so on - I got a new log file for every command, not only for errors or warnings.
My global `.npmrc` file has no extra settings.
Even if I add to it `loglevel=silent' I still have the new log files, but the info in the terminal is being cut off.

Any ideas to prevent these verbose log files being created and keep them only for errors and warning?

### Expected Behavior

Not to have a verbose log file for each npm command.

### Steps To Reproduce

_No response_

### Environment

- npm: 8.3.0
- Node: 17.3.0 installed with HomeBrew
- OS: MacOS
- platform: MacBook Pro
- npm config:

`npm config ls`:
```ini
; "builtin" config from /usr/local/lib/node_modules/npm/npmrc

prefix = "/usr/local"

; "user" config from /Users/(protected)/.npmrc
; node bin location = /usr/local/Cellar/node/17.3.0/bin/node
; cwd = /Users/(protected)
; HOME = /Users/(protected)
```

`npm config ls -l`:
```ini
; "default" config from default values

_auth = (protected)
access = null
all = false
allow-same-version = false
also = null
audit = true
audit-level = null
auth-type = "legacy"
before = null
bin-links = true
browser = null
ca = null
cache = "/Users/(protected)/.npm"
cache-max = null
cache-min = 0
cafile = null
call = ""
cert = null
ci-name = null
cidr = null
color = true
commit-hooks = true
depth = null
description = true
dev = false
diff = []
diff-dst-prefix = "b/"
diff-ignore-all-space = false
diff-name-only = false
diff-no-prefix = false
diff-src-prefix = "a/"
diff-text = false
diff-unified = 3
dry-run = false
editor = "vi"
engine-strict = false
fetch-retries = 2
fetch-retry-factor = 10
fetch-retry-maxtimeout = 60000
fetch-retry-mintimeout = 10000
fetch-timeout = 300000
force = false
foreground-scripts = false
format-package-lock = true
fund = true
git = "git"
git-tag-version = true
global = false
global-style = false
globalconfig = "/usr/local/etc/npmrc"
heading = "npm"
https-proxy = null
if-present = false
ignore-scripts = false
include = []
include-staged = false
include-workspace-root = false
init-author-email = ""
init-author-name = ""
init-author-url = ""
init-license = "ISC"
init-module = "/Users/(protected)/.npm-init.js"
init-version = "1.0.0"
init.author.email = ""
init.author.name = ""
init.author.url = ""
init.license = "ISC"
init.module = "/Users/(protected)/.npm-init.js"
init.version = "1.0.0"
json = false
key = null
legacy-bundling = false
legacy-peer-deps = false
link = false
local-address = null
location = "user"
lockfile-version = null
loglevel = "notice"
logs-max = 10
; long = false ; overridden by cli
maxsockets = 15
message = "%s"
metrics-registry = "https://registry.npmjs.org/"
node-options = null
node-version = "v17.3.0"
noproxy = [""]
npm-version = "8.3.0"
offline = false
omit = []
only = null
optional = null
otp = null
pack-destination = "."
package = []
package-lock = true
package-lock-only = false
parseable = false
prefer-offline = false
prefer-online = false
; prefix = "/usr/local/Cellar/node/17.3.0" ; overridden by builtin
preid = ""
production = null
progress = true
proxy = null
read-only = false
rebuild-bundle = true
registry = "https://registry.npmjs.org/"
save = true
save-bundle = false
save-dev = false
save-exact = false
save-optional = false
save-peer = false
save-prefix = "^"
save-prod = false
scope = ""
script-shell = null
searchexclude = ""
searchlimit = 20
searchopts = ""
searchstaleness = 900
shell = "/usr/local/bin/zsh"
shrinkwrap = true
sign-git-commit = false
sign-git-tag = false
sso-poll-frequency = 500
sso-type = "oauth"
strict-peer-deps = false
strict-ssl = true
tag = "latest"
tag-version-prefix = "v"
timing = false
tmp = "/var/folders/hr/bbbmspwn0pbbhplv4jpt6vcc0000gn/T"
umask = 0
unicode = true
update-notifier = true
usage = false
user-agent = "npm/{npm-version} node/{node-version} {platform} {arch} workspaces/{workspaces} {ci}"
userconfig = "/Users/(protected)/.npmrc"
version = false
versions = false
viewer = "man"
which = null
workspace = []
workspaces = null
yes = null

; "builtin" config from /usr/local/lib/node_modules/npm/npmrc

prefix = "/usr/local"

; "user" config from /Users/(protected)/.npmrc

; "cli" config from command line options

long = true
```

### LINKED ISSUE #4466: [feat] Custom logging location
# Before Opening Please...
- [X] Search for an existing/duplicate RRFC which might be relevant to your RRFC

[There is one RFC about logging: using STDERR for errors instead of STDOUT. This is unrelated.](npm/cli#4724)

## Motivation ("The Why")

Currently some CI providers, [specifically GitLab](https://docs.gitlab.com/ee/ci/pipelines/job_artifacts.html), restrict job artifacts to being relative to the repository. Presumably, this stops you from uploading runner files that may have dirty-state left from previous runs in non-containerized workflows. Specifically the GitLab documents say,

> The `paths` keyword determines which files to add to the job artifacts. **All paths to files and directories are relative** to the repository where the job was created.

This effectively means NPM logs on error can not be captured as job artifacts.

#### Current Behaviour

What I would like to do is something like this (ignore the actual error)

```
$ npx ng e2e
npm ERR! could not determine executable to run
npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-11-22T17_24_50_085Z-debug.lo
```

Then I want to upload `/root/.npm/_logs/2021-11-22T17_24_50_085Z-debug.log` as a job artifact. 

**This is not currently possible**

#### Desired Behaviour

In light of GitLab not permitting global files to be uploaded as job artifacts I would like an option to set the log file,

```sh
npx [--log <LOCATION>] ng e2e
```

Such that `LOCATION` could be relative to the cwd. 

Then I could do,

```
$ npx --log "./.npm/_logs/" ng e2e
```

And set `./.npm/_logs/*` as my artifacts to upload to GitLab.

### References

* https://github.com/npm/npm/issues/1548
* https://github.com/npm/npm/pull/5252
* https://github.com/npm/npm/pull/6744 (exact comment that explains this issue's importance to CI: I am not alone)
* https://stackoverflow.com/questions/21624030/change-npm-debug-log-location/23456148

### PR DESCRIPTION
**Pull request by lukekarrys**
~~Requires: https://github.com/npm/npm-registry-fetch/pull/107~~

This also allows logs-max to be set to 0 to disable log file writing.

Closes https://github.com/npm/cli/issues/4466
Closes https://github.com/npm/cli/issues/4206

### CODE REVIEW

File: lib/npm.js [hunk 1]
```diff
@@ -112,68 +115,60 @@ class Npm extends EventEmitter {
...
-        process.emit('timeEnd', `command:${cmd}`)
-      })
     } else {
-      return command.exec(args).finally(() => {
-        process.emit('timeEnd', `command:${cmd}`)
-      })
+      execPromise = command.exec(args)
     }
+
+    return execPromise.finally(timeEnd)
```
Comment (wraithgar) [inline lib/npm.js:154]: 👍

File: lib/npm.js [hunk 2]
```diff
@@ -329,6 +341,14 @@ class Npm extends EventEmitter {
     return this.#logFile.files
   }
 
+  get logsDir () {
+    return this.config.get('logs-dir') || join(this.cache, '_logs')
```
Comment (wraithgar) [inline lib/npm.js:336]: Can we default this like `cache`?

Comment (lukekarrys) [reply lib/npm.js:336]: I couldn't get that to work nicely with how we do config definitions. I tried setting the `defaults: join(cache, '_logs')` in the definitions, but then if you pass in `--cache=./my-cache`, that value isn't populated until after `config.load()`. So I don't think we can do that until we have a truly derived config system.

Comment (wraithgar) [reply lib/npm.js:336]: Ah right, that only works for the flattened versions.

**Review by wraithgar — APPROVED** [review]
Lots of good cleanup in this feat.

**Review by wraithgar — APPROVED** [review]

**Comment by lukekarrys** [comment]
Just a note that this PR should probably be rebased, as I added the deps as separate commits.

**Pull request merged**